python/ray/autoscaler/v2/instance_manager/node_provider.py

```diff
@@ -1,16 +1,160 @@
+    update_id: str
+
+
+@dataclass
+class CloudNodeProviderState:
+    """
+    The state of a cloud node provider.
+    """
+
+    # The cloud nodes that are currently running.
+    running: Dict[CloudInstanceId, CloudInstance] = field(default_factory=dict)
+    # Errors that have happened when launching nodes.
+    launch_errors: List[LaunchNodeError] = field(default_factory=list)
+    # Errors that have happened when terminating nodes.
+    termination_errors: List[TerminateNodeError] = field(default_factory=list)
+
+
+class ICloudNodeProvider(ABC):
+    """
+    The interface for a cloud node provider.
+
+    This interface is a minimal interface that should be implemented by the
+    various cloud node providers (e.g. AWS, and etc).
+
+    The cloud node provider is responsible for managing the cloud nodes in the
+    cluster. It provides the following main functionalities:
+        - Launch new cloud nodes.
+        - Terminate cloud nodes.
+        - Get the running cloud nodes in the cluster.
+
+    Below properties of the cloud node provider are assumed with this interface:
+
+    1. Eventually consistent
+    The cloud node provider is expected to be eventually consistent with the
+    cluster state. For example, when a node is request to be terminated/launched,
+    the node provider may not immediately reflect the change in its state.
+
+    2. Asynchronous
+    The node provider could also be asynchronous, where the termination/launch
+    request may not immediately return the result of the request.
+
+    3. Unique cloud node ids
+    Cloud node ids are expected to be unique across the cluster.
+
+    Usage:
+        ```
+            cloud_node_provider: ICloudNodeProvider = ...
+
+            # Update the cluster with a designed shape.
+            cloud_node_provider.update(
+                id="update_1",
+                target_running_nodes={
+                    "worker_nodes": 10,
+                    "ray_head": 1,
+                },
+                to_terminate=["node_1", "node_2"],
+            )
+
+            # Poll the state of the cloud node provider.
+            state = cloud_node_provider.get_state()
+
+            # Process the state of the cloud node provider.
+        ```
+    """
+
+    @abstractmethod
+    def get_state(self, errors_since_ns: Optional[int]) -> CloudNodeProviderState:
+        """Get the current state of the cloud node provider.
+
+        Args:
+            errors_since_ns: retrieve errors that happened since this timestamp.
+
+        Returns:
+            The current state of the cloud node provider.
+        """
+        pass
+
+    @abstractmethod
+    def update(
+        self,
+        id: str,
+        target_running_nodes: Dict[NodeType, int],
+        to_terminate: List[CloudInstanceId],
+    ) -> None:
+        """Update the cloud node provider state by launching
+         or terminating cloud nodes.
+
+        Args:
+            id: the id of the update request.
```

Comment:
should we make this id a monotonically increasing integer so that instead of `errors_since_ns` you can do `errors_since_update_id` so we avoid depending on the wall clock time?